platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,754 @@
+            sdmmc::HardwareOptions::SetBusWidth(width) => {
+                self.smhc.smhc_ctype.write(|w| match width {
+                    sdmmc::BusWidth::Single => w.card_wid().b1(),
+                    sdmmc::BusWidth::Quad => w.card_wid().b4(),
+                    sdmmc::BusWidth::Octo => w.card_wid().b8(),
+                });
+            }
+        }
+
+        let mut guard = self.isr.lock(self.smhc);
+        guard.data.op = match (params.kind, params.buffer) {
+            (sdmmc::CommandKind::Control, _) => SmhcOp::Control,
+            (sdmmc::CommandKind::Read(cnt), Some(buf)) => SmhcOp::Read {
+                buf,
+                cnt,
+                auto_stop,
+            },
+            (sdmmc::CommandKind::Write(cnt), Some(buf)) => SmhcOp::Write {
+                buf,
+                cnt,
+                auto_stop,
+            },
+            _ => {
+                tracing::error!("did not provide a buffer for read/write");
+                return Err(sdmmc::Error::Other);
+            }
+        };
+
+        // Statically allocate space for 16 DMA descriptors.
+        // Each descriptor can do a transfer of 4KB, giving a max total transfer of 64KB.
+        // By declaring them in this scope the descriptor memory will live long enough,
+        // however currently this is up to the user to guarantuee.
+        // Safety: a zero'ed descriptor is valid and will simply be ignored by the IDMAC.
+        let mut descriptors: [idmac::Descriptor; 16] = unsafe { core::mem::zeroed() };
+
+        // Perform checks on arguments to make sure we won't overflow the buffer
+        match &guard.data.op {
+            SmhcOp::Read { buf, cnt, .. } | SmhcOp::Write { buf, cnt, .. } => {
+                const DESCR_BUFF_SIZE: usize = 0x1000;
+
+                // Currently we limit the number of data that can be read at once
+                if *cnt > DESCR_BUFF_SIZE * descriptors.len() || buf.capacity() < *cnt {
+                    return Err(sdmmc::Error::Data);
+                }
+
+                tracing::debug!("Creating descriptor chain from buffer");
+                let buf_ptr = buf.as_slice().as_ptr();
+                let mut remaining = *cnt;
+                let mut index = 0;
+                while remaining > 0 {
+                    let buff_size = core::cmp::min(DESCR_BUFF_SIZE, remaining) as u16;
+                    remaining = remaining.saturating_sub(DESCR_BUFF_SIZE);
+                    let first = index == 0;
+                    let last = remaining == 0;
+                    descriptors[index] = idmac::Descriptor::try_from(idmac::DescriptorConfig {
+                        disable_int_on_complete: !last,
+                        first,
+                        last,
+                        buff_size,
+                        buff_addr: unsafe { buf_ptr.add(index * DESCR_BUFF_SIZE) },
+                        next_desc: (!last).then_some(&descriptors[index + 1] as *const _ as _),
```

Comment:
hmm, can this bounds check fail if we're on the last descriptor? should we maybe be using 
```suggestion
                        next_desc: (!last).then(|| &descriptors[index + 1] as *const _ as _),
```
so that the indexing only occurs if `last` is `false`?